soroban-env-host/src/vm.rs

```diff
@@ -72,160 +97,41 @@ impl std::hash::Hash for Vm {
-            // conservative we avoid even allowing this.
-            return Err(err!(
-                host,
-                (ScErrorType::WasmVm, ScErrorCode::InvalidInput),
-                "contract protocol number is newer than host",
-                got_proto
-            ));
-        }
-        Ok(())
-    }
-
-    fn check_meta_section(host: &Host, m: &Module) -> Result<u64, HostError> {
-        if let Some(env_meta) = Self::module_custom_section(m, meta::ENV_META_V0_SECTION_NAME) {
-            let mut limits = DEFAULT_XDR_RW_LIMITS;
-            limits.len = env_meta.len();
-            let mut cursor = Limited::new(Cursor::new(env_meta), limits);
-            if let Some(env_meta_entry) = ScEnvMetaEntry::read_xdr_iter(&mut cursor).next() {
-                let ScEnvMetaEntry::ScEnvMetaKindInterfaceVersion(v) =
-                    host.map_err(env_meta_entry)?;
-                Vm::check_contract_interface_version(host, v)?;
-                Ok(v)
-            } else {
-                Err(host.err(
-                    ScErrorType::WasmVm,
-                    ScErrorCode::InvalidInput,
-                    "contract missing environment interface version",
-                    &[],
-                ))
-            }
-        } else {
-            Err(host.err(
-                ScErrorType::WasmVm,
-                ScErrorCode::InvalidInput,
-                "contract missing metadata section",
-                &[],
-            ))
-        }
-    }
-
-    fn check_max_args(host: &Host, m: &Module) -> Result<(), HostError> {
-        for e in m.exports() {
-            match e.ty() {
-                wasmi::ExternType::Func(f) => {
-                    if f.params().len() > MAX_VM_ARGS || f.results().len() > MAX_VM_ARGS {
-                        return Err(host.err(
-                            ScErrorType::WasmVm,
-                            ScErrorCode::InvalidInput,
-                            "Too many arguments or results in wasm export",
-                            &[],
-                        ));
-                    }
-                }
-                _ => (),
-            }
-        }
-        Ok(())
+    #[cfg(feature = "testutils")]
+    pub fn get_all_host_functions() -> Vec<(&'static str, &'static str, u32)> {
+        HOST_FUNCTIONS
+            .iter()
+            .map(|hf| (hf.mod_str, hf.fn_str, hf.arity))
+            .collect()
     }
 
-    /// Instantiates a VM given the arguments provided in [`Self::new`]
+    /// Instantiates a VM given the arguments provided in [`Self::new`],
+    /// or [`Self::new_from_module_cache`]
     fn instantiate(
         host: &Host,
         contract_id: Hash,
-        module_wasm_code: &[u8],
+        parsed_module: Rc<ParsedModule>,
     ) -> Result<Rc<Self>, HostError> {
-        host.charge_budget(
-            ContractCostType::VmInstantiation,
-            Some(module_wasm_code.len() as u64),
-        )?;
+        let _span = tracy_span!("Vm::instantiate");
```

Comment:
It's hard to tell if the changes in here can't cause a divergence with the current protocol. It looks like a bunch of code was moved around and the ordering of possible errors has changed? Is the idea here that it doesn't matter where the error checks happen as long as they get done at some point because they aren't observable in the ledger?

Reply:
Yeah, the code motion is unfortunate. A few things have changed:

  - Parsing and instantiating are split up
  - The parsing code is moved to `parsed_module.rs`
  - Linking is only linking the necessary imports

As far as I know, all the same checks are happening, but those that were parse-related have moved with the parsing-related code and therefore _can_ happen earlier (in the new protocol). When replaying the old protocol I believe they _should_ all still happen at the same time and in the same order. I'll be testing this as much as I can in the next couple weeks.

Reply:
Yeah. All the test observations will also catch divergence.